allennlp/semparse/type_declarations/wikitables_variable_free.py

```diff
@@ -0,0 +1,126 @@
+# cells, and the executor can process them accordingly, but the grammar does not differentiate
+# between them. Also note that columns are basic types in this grammar.
+ROW_TYPE = NamedBasicType("ROW")
+# "LCOLUMN" to ensure the signature will be "l", to avoid confusion with cell. I decided to omit the
+# cell type eventually, and the decision to call this "LCOLUMN" is not relevant any more. But I'm
+# not changing it for now, because I might end up adding the cell type back later.
+COLUMN_TYPE = NamedBasicType("LCOLUMN")
+NUMBER_TYPE = NamedBasicType("NUMBER")
+DATE_TYPE = NamedBasicType("DATE")
+STRING_TYPE = NamedBasicType("STRING")
+
+BASIC_TYPES = {ROW_TYPE, COLUMN_TYPE, NUMBER_TYPE, DATE_TYPE, STRING_TYPE}
+STARTING_TYPES = {NUMBER_TYPE, DATE_TYPE, STRING_TYPE}
+
+# Complex types
+# Type for selecting the value in a column in a set of rows. "select" and "mode" functions.
+SELECT_TYPE = ComplexType(ROW_TYPE, ComplexType(COLUMN_TYPE, STRING_TYPE))
+
+# Type for filtering rows given a column. "argmax", "argmin" and "same_as" (select all rows with the
+# same value under the given column as the given row does under the given column)
+ROW_FILTER_WITH_COLUMN = ComplexType(ROW_TYPE, ComplexType(COLUMN_TYPE, ROW_TYPE))
+
+# "filter_number_greater", "filter_number_equals" etc.
+ROW_FILTER_WITH_COLUMN_AND_NUMBER = ComplexType(ROW_TYPE,
+                                                ComplexType(NUMBER_TYPE,
+                                                            ComplexType(COLUMN_TYPE, ROW_TYPE)))
+
+# "filter_date_greater", "filter_date_equals" etc.
+ROW_FILTER_WITH_COLUMN_AND_DATE = ComplexType(ROW_TYPE,
+                                              ComplexType(DATE_TYPE,
+                                                          ComplexType(COLUMN_TYPE, ROW_TYPE)))
+
+# "filter_in" and "filter_not_in"
+ROW_FILTER_WITH_COLUMN_AND_STRING = ComplexType(ROW_TYPE,
+                                                ComplexType(STRING_TYPE,
+                                                            ComplexType(COLUMN_TYPE, ROW_TYPE)))
+
+ROW_FILTER = ComplexType(ROW_TYPE, ROW_TYPE)  # first, last, previous, next etc.
+
+COUNT_TYPE = ComplexType(ROW_TYPE, NUMBER_TYPE)
+
+# Numerical operations on numbers in the given column. "max", "min", "sum", "average" etc.
+ROW_NUM_OP = ComplexType(ROW_TYPE, ComplexType(COLUMN_TYPE, NUMBER_TYPE))
+
+# Numerical difference within the given column.
+NUM_DIFF_WITH_COLUMN = ComplexType(ROW_TYPE, ComplexType(ROW_TYPE, ComplexType(COLUMN_TYPE,
+                                                                               NUMBER_TYPE)))
+
+# Date function takes three numbers and makes a date
+DATE_FUNCTION_TYPE = ComplexType(NUMBER_TYPE, ComplexType(NUMBER_TYPE, ComplexType(NUMBER_TYPE,
+                                                                                   DATE_TYPE)))
+
+
+COMMON_NAME_MAPPING = {}
+COMMON_TYPE_SIGNATURE = {}
+
+
+def add_common_name_with_type(name, mapping, type_signature):
```

Comment:
Type hints on the arguments here.  I'd also probably recommend just calling this `add_function`.

Reply:
Added type hints. The things being added to the dicts `COMMON_NAME_MAPPING` and `COMMON_TYPE_SIGNATURE` are not all functions (`all_rows` is a constant).